File: Leetcode/0100-same-tree/solution.go
package main

import "fmt"

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func isSameTree(p, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}
	if p == nil || q == nil {
		return false
	}
	if p.Val != q.Val {
		return false
	}
	return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

func buildTree(arr []interface{}) *TreeNode {
	if len(arr) == 0 || arr[0] == nil {
		return nil
	}
	root := &TreeNode{Val: arr[0].(int)}
	q := []*TreeNode{root}
	i := 1
	for len(q) > 0 && i < len(arr) {
		n := q[0]
		q = q[1:]
		if i < len(arr) && arr[i] != nil {
			n.Left = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Left)
		}
		i++
		if i < len(arr) && arr[i] != nil {
			n.Right = &TreeNode{Val: arr[i].(int)}
			q = append(q, n.Right)
		}
		i++
	}
	return root
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, exp bool) {
		if got == exp {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s got=%v\n", name, got)
			failed++
		}
	}
	cases := []struct {
		name string
		p, q []interface{}
		want bool
	}{
		{"Example 1", []interface{}{1, 2, 3}, []interface{}{1, 2, 3}, true},
		{"Example 2", []interface{}{1, 2}, []interface{}{1, nil, 2}, false},
		{"Both empty", []interface{}{}, []interface{}{}, true},
		{"One empty", []interface{}{}, []interface{}{1}, false},
		{"Different values", []interface{}{1, 2, 1}, []interface{}{1, 1, 2}, false},
		{"Single same", []interface{}{5}, []interface{}{5}, true},
		{"Single different", []interface{}{5}, []interface{}{6}, false},
	}
	for _, c := range cases {
		test(c.name, isSameTree(buildTree(c.p), buildTree(c.q)), c.want)
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0100-same-tree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        """Time O(min(|p|, |q|)), Space O(h)."""
        if not p and not q: return True
        if not p or not q: return False
        if p.val != q.val: return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)


def build_tree(arr):
    if not arr or arr[0] is None: return None
    root = TreeNode(arr[0])
    q = [root]; i = 1
    while q and i < len(arr):
        n = q.pop(0)
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    cases = [
        ("Example 1", [1, 2, 3], [1, 2, 3], True),
        ("Example 2", [1, 2], [1, None, 2], False),
        ("Both empty", [], [], True),
        ("One empty", [], [1], False),
        ("Different values", [1, 2, 1], [1, 1, 2], False),
        ("Single same", [5], [5], True),
        ("Single different", [5], [6], False),
    ]
    for n, p, q, exp in cases:
        got = sol.isSameTree(build_tree(p), build_tree(q))
        if got == exp: print(f"PASS: {n}"); passed += 1
        else: print(f"FAIL: {n} got={got}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
